### PR TITLE
Implement lexicon-based (with an aligned lexicon) mappings, add eng to eng-ipa

### PR DESCRIPTION
**Pull request by dhdaines**
Using Phonetisaurus, we can align existing dictionaries and use the resulting alignments to do g2p with index preservation.  So this does that for CMUdict.  There are some limitations:

- The aligned dictionary is quite large, especially once preprocessed and pickled (but loading it is quick now that we don't try to deepcopy it).
- We can and should store it more efficiently, like, by training G2P rules from it ;-)
- Alternate pronunciations aren't supported, because we have no concept of such things
- The "IPA" is somewhat approximate, because it is derived from the ARPABET.  Surely another dictionary could do better
- Obviously, a lexicon mapping will *not* handle anything that isn't in the lexicon, or punctuation, or spaces
- Sometimes the automatic alignments are a little strange

This PR also fixes a number of little problems with the Studio app and augments test coverage significantly.

This does actually add a mapping, so here's the checklist :)

- [ y] Mapping files are added in `g2p/mappings/langs`
- [ ]y Mapping is either added to an existing folder or a new folder has been added
- [ y] Language folder and files use appropriate [ISO 639-3 codes](https://en.wikipedia.org/wiki/List_of_ISO_639-3_codes)
- [ y] Config.yaml file includes all author names, and settings necessary
- [ y ] Please add some test data in `g2p/tests/public/data`. The added file should be a csv/tsv/psv file and each row should have the format `[input_mapping_code,output_mapping_code,input_string,output_string]`
- [ y] As the last step, G2P has been updated by running `g2p update` locally and committing the change
- [ y] You agree to license your contribution under the same license as this project (see [LICENSE](https://github.com/roedoejet/g2p/blob/main/LICENSE) file).

### CODE REVIEW

**Comment by dhdaines** [comment]
Just a note here that I may want to rethink putting the alignments inside `langs.pkl` - it makes it quite huge, and because somewhere in the code we `deepcopy` the mappings, it loads just as slowly as if we were to parse the text file.

**Comment by joanise** [comment]
Starting to review this PR now.

**Comment by dhdaines** [comment]
> Starting to review this PR now.

Thanks! This should be less controversial than #231  :)

File: g2p/mappings/langs/eng/README.md
Comment (joanise) [reply g2p/mappings/langs/eng/README.md:1]: This seems to be an upstream version of that work, but they don't have the exact file we're using: https://github.com/cmusphinx/cmudict

File: g2p/mappings/langs/eng/README.md
Comment (joanise) [inline g2p/mappings/langs/eng/README.md:7]: I like this tmp file, I generated it on my computer. So much easier to read the English dict in IPA than in ARPABET!

**Review by joanise — CHANGES_REQUESTED** [review]
This is really nice, thanks for adding it here! It'll be very useful. I have a bunch of comments, see them in the code, but one possible change request: how can we handle multiple word conversions?

If I do `g2p convert "Ceci est un test, Edouard." fra fra-ipa`, I get the expected output: `sʌsi ɛ œ̃ tɛ, ʌduaʁd.`
But if I do `g2p convert "This is a test, Edward." eng eng-ipa`, I just get a blank output.
In contrast, if I add the option `--tok` the English text is tokenized and I get the expected `ðɪs ɪz ʌ tɛst, ɛdwɜ˞d.`

I've held the position until now that the output should be the same with and without `--tok` and that discrepancies are in fact errors. The `--tok` was meant to show in the `--debugger` output the different tokens being handled one at a time.

But since English is lexicon based, that position doesn't really hang together anymore! Outputting an empty string, however, is really not great. Other cases pass through what they don't know how to handle, could lexicon-based g2p do the same?

Do we want them to systematically tokenize? Actually, no: Studio already tokenizes, we don't want to retokenize again on each token. I don't know what the best answer is here.

**Comment by joanise** [comment]
Question: could we compress the two `cmudict*.txt` files and read them with `zipfile.ZipFile`? That way, the PyPI bundle would be smaller. They each zip >75% to below one MB.

**Comment by joanise** [comment]
> Question: could we compress the two `cmudict*.txt` files and read them with `zipfile.ZipFile`? That way, the PyPI bundle would be smaller. They each zip >75% to below one MB.

Hmm, the whl files themselves are compressed, maybe this would not make any difference?

File: .github/workflows/studio-release-tests.yml
Comment (dhdaines) [reply .github/workflows/studio-release-tests.yml:38]: Thanks! I'm actually not certain that they are necessary, since codecov should combine the runs regardless, but we really aren't testing the app in the other tests.

File: g2p/app.py
Comment (dhdaines) [reply g2p/app.py:177]: Yes, exactly, this object isn't used for anything but creating JSON to send back to the client.

File: g2p/mappings/__init__.py
Comment (dhdaines) [reply g2p/mappings/__init__.py:407]: oops! I am also going to switch to using `%s` in the logging string to avoid unnecessary formatting

File: g2p/mappings/__init__.py
Comment (dhdaines) [reply g2p/mappings/__init__.py:413]: I think mypy or something complained about it, so I fixed it :)

File: g2p/mappings/langs/eng/README.md
Comment (dhdaines) [reply g2p/mappings/langs/eng/README.md:1]: There are a bunch of different versions of it floating around, unfortunately.  We *probably* should use the same one from the GitHub project as I think that's the authoritative one, but the phonesets are not always the same.

Comment (dhdaines) [reply g2p/mappings/langs/eng/README.md:1]: I switched to using the upstream version, and improved the docs (with a license)

File: g2p/tests/public/data/eng.csv
Comment (dhdaines) [reply g2p/tests/public/data/eng.csv:5]: Yes, I'm not entirely sure what the best answer is here, possibly making tokenization implicit with lexicon mappings, because in some cases they may need to segment words to look them up in the lexicon (for instance I think `'d`, `'ll`, and such are separate words in the English lexicon, and we should handle this even if we don't at the moment)

**Comment by joanise** [comment]
Well, there's a lesson for me there! `test_doctor_expensive.py` does not pass (and `g2p doctor` raises an exception) but you never noticed, because I disabled that test due to how slow it was. Clever of me, innit? It's only 13 seconds. I like how the g2p test suite is super quick, but maybe it's worth the extra 13 seconds anyway...

**Comment by joanise** [comment]
> should I re-enable the test?

Yeah, we should.

File: g2p/mappings/utils.py
Comment (joanise) [inline g2p/mappings/utils.py:395]: It's funny to me you bother with a default value when just `mapping.get("type") == "lexicon"` would be equivalent (`None` would also not equal `"lexicon"`).

**Review by joanise — COMMENTED** [review]
I think this is good now. If you can see how to make the langs.pkl even smaller, great, but not critical.
Letting Aidan review too before I mark this as approved.

**Comment by joanise** [comment]
And thanks for the cleaned up and rebased history!

File: g2p/tests/test_studio.py
Comment (roedoejet) [inline g2p/tests/test_studio.py:126]: a 5 second timeout? were you running into problems with the 1 second one? this might make the tests take quite a while. Did you try other numbers in between?

**Review by roedoejet — APPROVED** [review]
@dhdaines - this is so awesome! great work. I have a few things I would like to see. I'm fine to merge it as-is but if possible we would add some of these:

- documentation about how to make a mapping from a pronunciation dictionary. Maybe we could even add the make_alignments script logic into the `g2p` command line? I could see a blog post about this too.
- documentation about the different between lexicon/rule based mappings, particularly with respect to tokenization and the issue that @joanise highlighted.
- very minor clarification about whether the 5000ms timeout is needed in the test I commented on.

thanks so much @dhdaines !

**Review by roedoejet — CHANGES_REQUESTED** [review]
@dhdaines - this is so awesome! great work. I have a few things I would like to see. I'm fine to merge it as-is but if possible we would add some of these:

documentation about how to make a mapping from a pronunciation dictionary. Maybe we could even add the make_alignments script logic into the g2p command line? I could see a blog post about this too.
documentation about the different between lexicon/rule based mappings, particularly with respect to tokenization and the issue that @joanise highlighted.
very minor clarification about whether the 5000ms timeout is needed in the test I commented on.

thanks so much @dhdaines !

File: g2p/tests/test_studio.py
Comment (dhdaines) [reply g2p/tests/test_studio.py:126]: Ah, good catch, actually this is no longer needed now that I removed the deepcopy (I'm not sure that it actually waits 5 seconds, though)

**Comment by dhdaines** [comment]
> documentation about how to make a mapping from a pronunciation dictionary.  Maybe we could even add the make_alignments script logic into the g2p command line? I could see a blog post about this too

It's there in a readme, but it's very minimal - I will add at least something to `docs` before merging this.  I'm not certain that it can be done with tools easily installable from PyPI for the moment, which is the issue here.

> documentation about the different between lexicon/rule based mappings, particularly with respect to tokenization and the issue that @joanise highlighted.

Yes, agreed, this definitely needs to go somewhere in `docs`.  We *could* make tokenization always on for lexicon mappings if other things like convertextract depend on this, but I think it's best to encourage people to explicitly enable tokenization...

**Comment by dhdaines** [comment]
Hi! I think this is ready to merge, but I think you have to mark it as such (or not).  I would prefer to merge it, then do separate PRs for these two specific things:

- Integrate the alignment into the g2p CLI - I would prefer to do this at the same time as adding trainable rules, and also after removing `langs.pkl` from the git repository, to avoid a proliferation of large aligned dictionaries
- Find a solution for whitespace and punctuation with lexicon mappings - there's really no easy solution for this.  Real-world use cases should always use tokenization (and it seems that they do)

**Comment by joanise** [comment]
Suggestion: I've rebased this PR's branch to `dev.lexicon-compact` with a single commit of `langs.pkl` for the whole PR, thus minimizing the lasting size of these commits in the history.

`git diff dev.lexicon-compact dhdaines/dev.lexicon` confirms my rebased branch is identical to yours.

If you're OK with this, David, you can take my rebased history with yet another forced push:
```
git checkout dev.lexicon
git fetch origin
git diff origin/dev.lexicon-compact    # just for sanity checking, should output nothing
git reset --hard origin/dev.lexicon-compact  # switch to the version of the history with just one commit of langs.pkl
git push -f  # and force update the PR branch once more
```

**Review by joanise — APPROVED** [review]
This is ready to merge now, ideally with a merge commit, and ideally with my proposed rebased variant in branch `dev.lexicon-compact`.

**Review by roedoejet — APPROVED** [review]

**Pull request merged**
